### ISSUE TITLE
Inconsistent behavior in version 2.0.4 and versions after 2.0.4

### ISSUE THREAD

**Issue by shuyangdu** [Microsoft/LightGBM#974]
I used lightgbm for a regression task. Here are the parameters I used:
`params_gbd_l1 = {
    'max_bin': 80,
    'learning_rate': 0.0116,
    'boosting_type': 'gbdt',
    'objective': 'regression_l1',
    'feature_fraction': 0.94,
    'bagging_fraction': 0.85,
    'bagging_freq': 80,
    'num_leaves': 110,
    'lambda_l2': 86.9,
    'n_estimators': 450,
}`
I used the exact same data set and 5-fold cross validation index. For version 2.0.4 I can get average test MAE around 0.068 but for version after 2.0.4 I can only get MAE around 0.072. I have also done the hyper parameter search for using different versions, and the difference between test MAE is pretty the same as above. 

And this should not be an issue of random seed since using the same feature, I can reproduce exact the same MAE.

I have also tried to specify categorical features or not, and the difference remains the same.

Have we changed any vital implementation after version 2.0.4? Does this have something to do with the random forest mode we added after 2.0.4?

## Environment info
Operating System: macOS Sierra 10.12.5
CPU: 2.6 GHz Intel Core i7
C++/Python/R version: Python 2.7

**Comment by guolinke** [comment]
@shuyangdu 
after 2.0.4, we change the missing value handling, which use the ```na``` as missing value (previous is zero).
you can try to set ```zero_as_missing``` to use zero to represent missing.

**Comment by shuyangdu** [comment]
Thanks for your response, Guolin! I have tried to set zero_as_missing to true but the difference is still the same. Is there anything else that may affect the model behavior?

**Comment by guolinke** [comment]
@shuyangdu you can try the bigger ```max_bin```, we also modify this algorithm.

**Comment by shuyangdu** [comment]
Actually, I have tried to use different parameters (including bigger `max_bin`) in my hyper-parameter optimization and all of the results have similar differences in test MAE. Any other factors could cause this? And would it be any problems if I continue to use 2.0.4 (maybe you fixed some bugs in later versions)?
